src/backend/distributed/master/master_delete_protocol.c

```diff
@@ -163,7 +167,208 @@ master_apply_delete_command(PG_FUNCTION_ARGS)
 
 
 /*
- * master_drop_shards attempts to drop all shards for a given relation.
+ * master_delete_from_shards takes in a delete query string and pushes the
+ * query to shards. It finds shards that match the criteria defined in the
+ * delete command, generates the same delete query string for each of the
+ * found shards with distributed table name replaced with the shard name and
+ * sends the queries to the workers. It uses one-phase or two-phase commit
+ * transactions depending on citus.copy_transaction_manager value.
+ */
+Datum
+master_delete_from_shards(PG_FUNCTION_ARGS)
+{
+	text *queryText = PG_GETARG_TEXT_P(0);
+	char *queryString = text_to_cstring(queryText);
+	bool isTopLevel = true;
+	DeleteStmt *deleteStatement = NULL;
+	bool failOK = false;
+	Oid relationId = InvalidOid;
+	List *queryTreeList = NIL;
+	Query *deleteQuery = NULL;
+	Node *whereClause = NULL;
+	Node *deleteCriteria = NULL;
+	LOCKTAG lockTag;
+	bool sessionLock = false;
+	bool dontWait = false;
+	List *shardIntervalList = NIL;
+	Index tableId = 1;
+	List *restrictClauseList = NIL;
+	List *prunedShardIntervalList = NIL;
+	ListCell *shardIntervalCell = NULL;
+	List *connectionList = NIL;
+	int32 affectedTupleCount = 0;
+	List *tableFailedPlacementList = NIL;
+	ListCell *tableFailedPlacementCell = NULL;
+	Node *queryTreeNode;
+
+	PreventTransactionChain(isTopLevel, "master_delete_from_shards");
+
+	queryTreeNode = ParseTreeNode(queryString);
+	if (!IsA(queryTreeNode, DeleteStmt))
+	{
+		ereport(ERROR, (errmsg("query \"%s\" is not a delete statement",
+							   queryString)));
+	}
+
+	deleteStatement = (DeleteStmt *) queryTreeNode;
+	relationId = RangeVarGetRelid(deleteStatement->relation, NoLock, failOK);
+
+	CheckDistributedTable(relationId);
+
+	queryTreeList = pg_analyze_and_rewrite(queryTreeNode, queryString, NULL, 0);
+	deleteQuery = (Query *) linitial(queryTreeList);
+	CheckTableCount(deleteQuery);
+
+	/* get where clause and flatten it */
+	whereClause = (Node *) deleteQuery->jointree->quals;
+	deleteCriteria = eval_const_expressions(NULL, whereClause);
+	CheckDeleteCriteria(deleteCriteria);
+
+	SET_LOCKTAG_ADVISORY(lockTag, MyDatabaseId, relationId, 0, 0);
+	LockAcquire(&lockTag, ExclusiveLock, sessionLock, dontWait);
```

Comment:
LockAllShards function takes a `shardintervallist`. Why can't I use it with the `prunedShardIntervalList`? Is it possible that the name of the function is not all correct? (`LockShards`, instead of `LockAllShards`).